Fix #288 Biomeの推奨Lintルールを適用する

meganetaaan opened the pull request:
https://github.com/stack-chan/stack-chan/issues/288 への対応です。



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **Bug Fixes**
	- Enhanced strict equality checks across various modules to prevent type coercion issues.
	- Improved error handling in the HTTP server service and preference server.

- **New Features**
	- Introduced a new type alias for preference values, enhancing type safety.
	- Improved handling of input and output paths in speech generation scripts for better clarity.
	- Updated return types for `onLaunch` and `onRobotCreated` properties in the StackchanMod interface for clearer expectations.
	- Added a new `SimButton` class for improved button functionality in the simulator environment.

- **Documentation**
	- Updated type imports in `mp3streamer` and `wavstreamer` modules for clarity on type usage.
	- Enhanced logging and error messages in various scripts for better debugging.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

meganetaaan committed:
enable noForEach

meganetaaan committed:
enable noUselessConstructor

meganetaaan committed:
enable noDoubleEquals

meganetaaan committed:
fix lint

meganetaaan committed:
Fix noDoubleEquals

sskw-ugo commented on firmware/stackchan/dialogues/dialogue-gemini.ts:
```diff
@@ -84,7 +84,7 @@ export class GeminiDialogue {
       .map((c) => ({
         parts: [{ text: c.content }],
-        role: c.role == 'assistant' ? 'model' : 'user',
+        role: c.role === 'assistant' ? 'model' : 'user',
```
nullチェックでtriple equalsを使うとundefinedにマッチしなくなるため目的を果たせません。したがって指摘の箇所はこのままでOKです。

meganetaaan committed:
Apply lint/format to `scripts`

meganetaaan committed:
Fix noUselessElse

meganetaaan committed:
Fix noParameterReassing

meganetaaan committed:
fix noConfusingVoidType

meganetaaan committed:
Fix useExponentiationOperator

meganetaaan committed:
Fix useSingleVarDeclarator

meganetaaan committed:
Fix useDefaultParameterLast

meganetaaan committed:
typo

meganetaaan committed:
Fix noAssignInExpressions

meganetaaan committed:
Fix noImplicitAnyLet

meganetaaan committed:
Fix noExplicitAny

meganetaaan committed:
import HTTPClient

meganetaaan committed:
ignore type error of Headers

meganetaaan commented on firmware/stackchan/drivers/dynamixel.ts:
```diff
@@ -451,7 +451,7 @@ class Dynamixel {
   async readFirmwareVersion(): Promise<Maybe<{ version: number }>> {
     const values = await this.#sendCommand(INSTRUCTION.READ, ADDRESS.VERSION_OF_FIRMWARE, 1)
-    if (values != null && values[1] == 0) {
+    if (values != null && values[1] === 0) {
```
`values[1]`はステータスコードを表しています。`values[1] === 0`はコマンド成功を意味しており、意図した通りの実装です。

meganetaaan commented:
biome対応をしました。@stc1988 レビューをお願いできますでしょうか :pray:
一部Moddable側の型の修正が必要な箇所があり、 `@ts-ignore`アノテーションを付けて一時的に型を無視しています。
https://github.com/Moddable-OpenSource/moddable/pull/1420

meganetaaan commented:
🐰さんから追加の指摘を受けたので確認、修正します

sskw-ugo commented on firmware/package.json:
```diff
@@ -4,9 +4,9 @@
   "scripts": {
-    "format": "biome format stackchan mods tests",
+    "format": "biome format",
     "format:fix": "npm run format -- --write",
-    "lint": "biome lint stackchan mods tests",
+    "lint": "biome lint",
```
- `@biomejs/biome` がインストールされています
- LintやFormatの影響を目視でレビューしています

meganetaaan committed:
void -> undefined for consistency

sskw-ugo commented on firmware/stackchan/default-mods/mod.ts:
```diff
@@ -3,7 +3,7 @@ import { onLaunch } from 'default-mods/on-launch'
 export interface StackchanMod {
-  onLaunch?: () => Promise<void | boolean> | void | boolean
+  onLaunch?: () => Promise<undefined | boolean> | undefined | boolean
   onRobotCreated?: (robot: Robot, option?: unknown) => Promise<void> | void
```
修正しました

meganetaaan committed:
void -> undefined

meganetaaan committed:
temporary (unsafe) cast instead of @ts-ignore

meganetaaan committed:
format

meganetaaan committed:
mod.onLaunchの型を変更

- 「booleanまたはundefined」は紛らわしいため、常にbooleanを返す
- 戻り値がtrueならrobot起動、falseなら起動しない
- modでonLaunchが戻り値を返さない場合の挙動が変わることに注意（そこでｽﾀｯｸﾁｬﾝの起動が中断される）

meganetaaan commented on firmware/stackchan/default-mods/mod.ts:
```diff
@@ -3,8 +3,8 @@ import { onLaunch } from 'default-mods/on-launch'
-  onLaunch?: () => Promise<void | boolean> | void | boolean
-  onRobotCreated?: (robot: Robot, option?: unknown) => Promise<void> | void
+  onLaunch?: () => Promise<undefined | boolean> | undefined | boolean
+  onRobotCreated?: (robot: Robot, option?: unknown) => Promise<undefined> | void
```
定義を見直して以下のようにしました

```
export interface StackchanMod {
  onLaunch?: () => Promise<boolean> | boolean
  onRobotCreated?: (robot: Robot, option?: unknown) => Promise<void> | void
}
```

`onRobotCreated`の戻り値は使われないためvoidが適切です。
一方`onLaunch`は「booleanが返された場合、falseならrobotを起動しない。undefinedの場合は起動する」という挙動でした。falseと（通常falsyである）undefinedとで結果が異なるのは一貫性がありません。
そこで、onLaunchは必ずbooleanを返すよう型で規定します。

meganetaaan commented:
@stc1988 ready for reviewです！
1点共有として、biomeの型の指摘に対応するために破壊的変更を加えました。
https://github.com/stack-chan/stack-chan/pull/298#discussion_r1789024177

stc1988 reviewed (approved):
@meganetaaan 確認しましたLGTMです 🙆‍♂️
